src/test/java/com/fasterxml/jackson/databind/util/concurrentlinkedhashmap/ConcurrentLinkedHashMapStressTest.java

```diff
@@ -53,8 +51,9 @@ public void testManyEntries() throws Exception {
             executor.shutdown();
         }
         executor.awaitTermination(waitSeconds, TimeUnit.SECONDS);
-        //clhm.executor.shutdown();
-        //clhm.executor.awaitTermination(waitSeconds, TimeUnit.SECONDS);
+
+        //ConcurrentLinkedHashMap runs evictions in the background using its own ExecutorService
+        //the following code retries the assertions to allow for the evictions to finish
```

Comment:
This isn't actually true. You are likely running into overflows due to the calls being amortized across threads, where the write buffer is unbounded and a stress test writes w/o doing any work. In a [memory leak stress test](https://github.com/ben-manes/concurrentlinkedhashmap/blob/master/src/test/java/com/googlecode/concurrentlinkedhashmap/MemoryLeakTest.java) this stayed within tolerance, but the cache is allowed to temporarily exceed its maximum. In Caffeine this write buffer is bounded to induce back pressure so that eviction can keep up. Regardless, the only way to keep strictly to a maximum size is for writes to be serialized on an exclusive lock, whereas allowing some temporary overflow for concurrent writes is more desirable.

Reply:
Thanks @ben-manes for the clarification - I'm happy for the CLHM to overflow occasionally under load.

The retries for the assertions helps to get them to pass. In the earlier versions of the test, the assertions failed on map size and on some of the get comparisons, despite the test having a wait that waits for the async tasks from the executor service to complete.

Reply:
Ok. And just to make clear -- this Map implementation does not use or rely on background threads, but just on non-sync updates via regular usage? That's fine, but I definitely would not want to have extra threads for such usage.

Reply:
Yes, it piggybacks on the caller to run maintenance work, like `LinkedHashMap`. It differs by instead of performing the LRU reordering immediately under a lock (a few pointer changes), it captures those access events into cheap buffers. Then when a buffer is full or a write occurs, it uses a `tryLock` to replay the LRU reordering operations. This way threads can defer that work to whomever drains, which is cheaper than lock contention. This can lead to some events being queue'd up after a drain but before the maintenance lock is released, causing it to wait until the next activity to flush those out.

I had forgotten in v1.2 there was temporarily an executor option (since removed) to flush these and Guava/Caffeine added `Cache.cleanUp()` instead. Caffeine does offer an executor setting, but that is because while the algorithms are very cheap the callouts to user code (e.g. an eviction listener) may not be. In that case we default async via `ForkJoinPool.commonPool()` to favor not running it on an unsuspecting calling thread. However many just use `Runnable::run` for a same-thread call to mimic CLHM / Guava.

Reply:
The ExecutorService in CLHM is now gone